Create an evaluator in LangSmith UI from the SDK (#4514)
langchain-ai/docs: src/langsmith/manage-evaluators-sdk.mdx

---
title: Manage evaluators with the SDK
sidebarTitle: Manage evaluators programmatically
description: Create, retrieve, update, list, and delete LangSmith evaluators programmatically with the SDK.
icon: "code"
tag: "SDK"
---

Use the LangSmith SDK to create and manage [evaluators](/langsmith/evaluation-concepts#evaluators) programmatically. Evaluators created through the SDK are [workspace-level](/langsmith/administration-overview#workspaces) resources that appear in the **Evaluators** table in the [LangSmith UI](https://smith.langchain.com), the same as [evaluators created in the UI](/langsmith/evaluators#create-an-evaluator-in-the-ui). You can attach them to datasets to run [offline evaluations](/langsmith/evaluation-concepts#offline-evaluations) and to tracing projects to run [online evaluations](/langsmith/evaluation-concepts#online-evaluations). Use the SDK to automate evaluator management and integrate evaluation into your existing workflows.

<Note>

> This guide uses the Python and TypeScript SDKs. For installation and setup, refer to the [Python SDK documentation](https://reference.langchain.com/python/langsmith) and [TypeScript SDK documentation](https://reference.langchain.com/javascript/modules/langsmith.html).

Reviewer: This note could probably move down into the prerequisites to be on hand when/if the user needs to set up the SDK.

Author: good suggestion! updated

</Note>

## Prerequisites

Managing evaluators through the SDK requires:

- Python: `langsmith>=0.8.17` (PyPI)
- TypeScript: `langsmith>=0.7.10` (npm)

> The examples below initialize the client with no arguments, so it reads the `LANGSMITH_API_KEY` and `LANGSMITH_ENDPOINT` environment variables. Configure your [API key](/langsmith/create-account-api-key) through environment variables rather than hardcoding it.

Reviewer: ```suggestion
The examples on this page initialize the client with no arguments, so it reads the `LANGSMITH_API_KEY` and `LANGSMITH_ENDPOINT` environment variables. Configure your [API key](/langsmith/create-account-api-key) through environment variables rather than hardcoding it.
```

Author: updated

> In the examples below, replace placeholders such as `<evaluator-uuid>` with the corresponding information from LangSmith.

Reviewer: ```suggestion
In the following examples, replace placeholders such as `<evaluator-uuid>` with the corresponding information from LangSmith.
```

Author: updated

## Create an evaluator

### Code evaluator

> 

Reviewer: Anything we want to add here to introduce the code snippet?

Author: added a sentence

</Note>

<CodeGroup>
```python Python
created = client.online_evaluators.create(
    name="LLM judge",
    type="llm",
    llm_evaluator={
        "prompt_repo_handle": "<prompt-repo-handle>",
        "commit_hash_or_tag": "<commit-hash-or-tag>",
        "variable_mapping": {
            "input": "inputs.question",
            "output": "outputs.answer",
            "reference": "reference.answer",
        },
    },
)
evaluator_id = created.evaluator.id
```
```typescript TypeScript
const created = await client.onlineEvaluators.create({
  name: "LLM judge",
  type: "llm",
  llm_evaluator: {
    prompt_repo_handle: "<prompt-repo-handle>",
    commit_hash_or_tag: "<commit-hash-or-tag>",
    variable_mapping: {
      input: "inputs.question",
      output: "outputs.answer",
      reference: "reference.answer",
    },
  },
});
const evaluatorId = created.evaluator?.id;
```
</CodeGroup>

The `prompt_repo_handle` is the prompt's internal repository name, not its display title or URL. To find it, list your workspace prompts and read the `repo_handle` field, or retrieve a specific prompt by its identifier in LangSmith. The identifier of a prompt can be in the format:

> - `promptName` (for private prompts), for example `my-prompt`.

Reviewer 2: ooc this is an interesting note - do they need to know this? it may be almost too specific and confusing

Author: I think this is a necessary clarification to confirm what is the actual `repo_handle` to pass to `client.onlineEvaluators.create`.

    llm_evaluator_id,
    llm_evaluator={
        "prompt_repo_handle": "<prompt-repo-handle>",
        "commit_hash_or_tag": "<commit-hash-or-tag>",
        "variable_mapping": {
            "input": "inputs.question",
            "output": "outputs.answer",
        },
        "use_corrections_dataset": True,
        "num_few_shot_examples": 3,
    },
)
```
```typescript TypeScript
const llmEvaluatorId = "<llm-evaluator-uuid>";

await client.onlineEvaluators.update(llmEvaluatorId, {
  llm_evaluator: {
    prompt_repo_handle: "<prompt-repo-handle>",
    commit_hash_or_tag: "<commit-hash-or-tag>",
    variable_mapping: {
      input: "inputs.question",
      output: "outputs.answer",
    },
    use_corrections_dataset: true,
    num_few_shot_examples: 3,
  },
});
```
</CodeGroup>

## List evaluators

Filter by name, type, feedback key, attached resource, or tag value, and sort or paginate the results. `list()` auto-paginates through every match when you iterate the returned object directly. `limit` sets the per-request page size (1 to 100), not the total number of results. `sort_by` is optional, accepts `created_at` or `updated_at`, and defaults to `created_at`.

> 

Reviewer 2: nice!

Author: thanks

    type="llm",
    sort_by="updated_at",  # "created_at" (default) or "updated_at"
    sort_by_desc=False,
    limit=20,
    offset=0,
)
```
```typescript TypeScript
// Read a single page of results
const page = await client.onlineEvaluators.list({
  name_contains: "correctness",
  type: "code",
  limit: 10,
});
for (const evaluator of page.evaluators) {
  console.log(evaluator.id, evaluator.name, evaluator.type);
}

// Collect every match into an array
const evaluators = [];
for await (const evaluator of client.onlineEvaluators.list({
  feedback_key: "correctness",
  limit: 20,
})) {
  evaluators.push(evaluator);
}

// Filter, sort, and paginate
await client.onlineEvaluators.list({
  feedback_key: "correctness",
  name_contains: "judge",
  resource_id: ["<project-or-dataset-uuid>"],
  tag_value_id: ["<tag-value-uuid>"],
  type: "llm",
  sort_by: "updated_at", // "created_at" (default) or "updated_at"
  sort_by_desc: false,
  limit: 20,
  offset: 0,
});
```
</CodeGroup>

## Track evaluator spend

Retrieve estimated USD spend and trace counts for your evaluators:

- **`period_start`**: A date-only ISO string, such as `2026-06-29`. Passing a datetime returns a 400 error.

> - **Window**: `period_start` starts a fixed 7-day window that includes `period_start` and the six days after it. The window is half-open, `[period_start, period_start + 7 days)`, so `period_end` (`period_start` plus 7 days) is excluded. For example, a `period_start` of `2026-06-29` covers `2026-06-29` through `2026-07-05`, and `2026-07-06` is excluded.

Reviewer: should this be indented below `period_start` because it's not another field?

Author: good suggestion! It bothered me, but I couldn't think of how to fix it.

});
for (const group of spend.groups ?? []) {
  console.log(group.evaluator_name, group.total_spend_usd, group.total_trace_count);
}

// Group spend by evaluator
const spendByEvaluator = await client.onlineEvaluators.spend({
  period_start: startDate,
  group_by: "evaluator",
  type: "llm",
});
console.log("Group by evaluator");
for (const group of spendByEvaluator.groups ?? []) {
  console.log(group.evaluator_name, group.total_spend_usd, group.total_trace_count);
}

// Group spend by resource
const spendByResource = await client.onlineEvaluators.spend({
  period_start: startDate,
  group_by: "resource",
  type: "llm",
});
console.log("Group by resource");
for (const group of spendByResource.groups ?? []) {
  console.log(group.session_name, group.dataset_name, group.total_spend_usd, group.total_trace_count);
}

// Group spend by run_rule
const spendByRunRule = await client.onlineEvaluators.spend({
  period_start: startDate,
  group_by: "run_rule",
  type: "llm",
});
console.log("Group by run_rule");
for (const group of spendByRunRule.groups ?? []) {
  console.log(group.run_rule_name, group.total_spend_usd, group.total_trace_count);
}
```
</CodeGroup>

## Delete an evaluator

> An evaluator cannot be deleted while it is attached to a tracing project or dataset. Set `delete_run_rules` to `true` to delete the run rules that reference the evaluator before deleting the evaluator.

Reviewer: ```suggestion
You cannot delete an evaluator while it is attached to a tracing project or dataset. Set `delete_run_rules` to `true` to delete the run rules that reference the evaluator before deleting the evaluator.
```

Author: updated

<CodeGroup>
```python Python
evaluator_id = "<evaluator-uuid>"

client.online_evaluators.delete(
    evaluator_id,
    delete_run_rules=True, # run rules referencing the evaluator are deleted first
)
```
```typescript TypeScript
const evaluatorId = "<evaluator-uuid>";

await client.onlineEvaluators.delete(evaluatorId, {
  delete_run_rules: true, // run rules referencing the evaluator are deleted first
});
```
</CodeGroup>

## Related

- [Manage evaluators](/langsmith/evaluators): View and manage evaluators in the LangSmith UI.

> - [Set up LLM-as-a-judge online evaluators](/langsmith/online-evaluations-llm-as-judge)

Reviewer 2: Nit - Maybe make it more clear that the /langsmith/online-evaluations-llm-as-judge page is only for the UI.

Author: updated
